store/index/index.go

```diff
@@ -90,119 +92,226 @@ const BucketPoolSize = 1024
+	if err != nil {
+		return nil, err
+	}
+
+	err = upgradeIndex(path, headerPath)
+	if err != nil {
+		return nil, err
+	}
+
+	header, err := readHeader(headerPath)
+	if os.IsNotExist(err) {
+		header = newHeader(indexSizeBits)
+		if err = writeHeader(headerPath, header); err != nil {
 			return nil, err
 		}
-		sizeBuckets, err = NewSizeBuckets(indexSizeBits)
+		lastIndexPath = indexFileName(path, 0)
+		file, err = openNewFileAppend(lastIndexPath)
 		if err != nil {
 			return nil, err
 		}
 	} else {
 		if err != nil {
 			return nil, err
 		}
-		buckets, sizeBuckets, err = scanIndex(path, indexSizeBits)
+
+		if header.BucketsBits != indexSizeBits {
+			return nil, types.ErrIndexWrongBitSize{header.BucketsBits, indexSizeBits}
+		}
+
+		lastIndexNum, err = scanIndex(path, header.FirstFile, indexSizeBits, buckets, sizeBuckets, fileBuckets)
 		if err != nil {
 			return nil, err
 		}
-		file, err = openFileRandom(path, os.O_RDWR|os.O_APPEND|os.O_EXCL)
+
+		lastIndexPath = indexFileName(path, lastIndexNum)
+		file, err = openFileAppend(lastIndexPath)
 		if err != nil {
 			return nil, err
 		}
-		length = types.Position(stat.Size())
-	}
-	return &Index{
-		indexSizeBits,
-		buckets,
-		sizeBuckets,
-		file,
-		bufio.NewWriterSize(file, indexBufferSize),
-		primary,
-		sync.RWMutex{},
-		0,
-		make(bucketPool, BucketPoolSize),
-		make(bucketPool, BucketPoolSize),
-		length,
-	}, nil
-}
-
-func scanIndex(path string, indexSizeBits uint8) (Buckets, SizeBuckets, error) {
-	// this is a single sequential read across the whole index
-	file, err := openFileForScan(path)
-	if err != nil {
-		return nil, nil, err
 	}
-	defer func() {
-		_ = file.Close()
-	}()
-	header, bytesRead, err := ReadHeader(file)
+
+	fi, err := file.Stat()
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
-	if header.BucketsBits != indexSizeBits {
-		return nil, nil, types.ErrIndexWrongBitSize{header.BucketsBits, indexSizeBits}
+
+	idx := &Index{
+		sizeBits:    indexSizeBits,
+		buckets:     buckets,
+		sizeBuckets: sizeBuckets,
+		fileBuckets: fileBuckets,
+		file:        file,
+		fileNum:     lastIndexNum,
+		headerPath:  headerPath,
+		writer:      bufio.NewWriterSize(file, indexBufferSize),
+		Primary:     primary,
+		bucketLk:    sync.RWMutex{},
+		curPool:     make(bucketPool, BucketPoolSize),
+		nextPool:    make(bucketPool, BucketPoolSize),
+		length:      types.Position(fi.Size()),
+		basePath:    path,
+		updateSig:   make(chan struct{}, 1),
+		gcDone:      make(chan struct{}),
 	}
-	buckets, err := NewBuckets(indexSizeBits)
+
+	go idx.garbageCollector()
+
+	return idx, nil
+}
+
+func indexFileName(basePath string, fileNum uint32) string {
+	return fmt.Sprintf("%s.%d", basePath, fileNum)
+}
+
+func scanIndex(basePath string, fileNum uint32, indexSizeBits uint8, buckets Buckets, sizeBuckets SizeBuckets, fileBuckets FileBuckets) (uint32, error) {
+	if len(buckets) != 1<<indexSizeBits {
+		panic("wrong size of buckets")
```

Comment:
I am curious; why panic instead of returning an error?

Reply:
It is a programming error and not a result of any misconfiguration or data problem.  It can only happen if the code got messed up and the value of `indexSizeBits`, or the returns from `NewBuckets` or from `NewSizeBuckets` were accidentally changed in the code.

Reply:
Actually, not needed at all, especially since `indexSizeBits` is not even used.